### PR TITLE
WIP: Add support for IAsyncEnumerable (#1320)

### PR DESCRIPTION
**Pull request by kuzn-ilya**

### CODE REVIEW

**Comment by Shane32** [comment]
This is a pretty good implementation of basic support of an `IAsyncEnumerable`.  It validates nodes as they are being received and it does not require a `List<T>` to be created prior to the `List<ExecutionNode>` being created.  But that's about it.  Child nodes still cannot execute until the complete list has been received, so the performance increase is small at best.  Due to other issues (see below), I would guess that it runs slower for small result sets, and slightly better for very large result sets.  Certainly a minor difference either way.

Let's consider some other issues.  First, using reflection to examine supported interfaces is relatively slow, and as coded here, creates use of LINQ on the 'hot path' causing extra allocations and execution time.  `IAsyncEnumerable` support could be configured to be a fallback if `IEnumerable` is not matched, but this would not work for linq2db whose `IQueryable` objects support both `IAsyncEnumerable<T>` and `IEnumerable` interfaces.

Second, constructing a generic method is extremely slow -- a few orders of magnitude.  A concurrent dictionary should be used to cache these delegates.  Even after the method is cached, this does cause a dictionary lookup, which consumes a measurable amount of time, even if small.  This is just yet another reason why using `ToListAsync` before the call to GraphQL.NET will almost certainly perform better than this PR for small data sets.

Third, public API method signatures changed, so it could not be implemented until GraphQL.NET v5.

To solve some of these issues, I would investigate adding an interface and/or wrapper for `IAsyncEnumerable`s to make it more of an opt-in scenario.  A wrapper is a single allocation, but one is certainly less than the many allocations currently needed to check for `IAsyncEnumerable` support of an object.  So then the execution code could be something like `return is IAsyncResult resultAsync`.  Field builders can detect `IAsyncEnumerable` types (when strongly typed) and automatically use the correct field resolver and/or wrapper.  The wrapper can implement `IAsyncEnumerable<object>` so that there is no need to use reflection during execution.

Note that all of the above is true regardless of support of `IAsyncEnumerable` via `Microsoft.Bcl.AsyncInterfaces` for .NET Standard 2.0.  As of now, I still do not think it would be worth adding support to .NET Standard 2.0 projects at the cost of the extra dependency.  There is so little to be gained even if all the issues above were fixed and/or optimized.  Any new projects would likely be built on .NET Standard 2.1 anyway (which is .NET Core 3.0+).  However, as a counter-argument, it could also be said that old projects would likely continue using an old version of GraphQL.NET, and so it would not matter if `Microsoft.Bcl.AsyncInterfaces` was required for .NET Standard 2.0 within GraphQL.NET v5.

In conclusion, I would not recommend merging this PR (for v5) until benchmarks can confirm that there is no/negligible speed and memory difference for non-async queries.  I would also expect that we should investigate revising the execution strategy to start executing nodes as they are received from the source before making this change.

**Comment by sungam3r** [comment]
Agree with @Shane32 . In general, the support of `IAsyncEnumerable` is not so easy. In addition, if we go in this direction, then sure without the use of reflection and LINQ. I close this PR because there was no feedback for the last comment for almost a year.
